packages/discount-code-importer/test/main.spec.js

```diff
@@ -1,11 +1,247 @@
+    warn: () => {},
+    info: () => {},
+    verbose: () => {},
+  }
+
+  const codes = JSON.parse(
+    fs.readFileSync(path.join(__dirname, 'helpers/sampleCodes.json')),
+  )
+
+  let codeImport
+  beforeEach(() => {
+    codeImport = new DiscountCodeImport(logger, {
+      apiConfig: {
+        projectKey: 'asafaelhn',
+      },
+    })
+  })
+
+  it('should be a function', () => {
+    expect(typeof DiscountCodeImport).toBe('function')
+  })
+
+  it('should set default properties', () => {
+    expect(codeImport.logger).toEqual(logger)
+    expect(codeImport.client).toBeDefined()
+    expect(codeImport.apiConfig).toBeDefined()
+    expect(codeImport.batchSize).toBeDefined()
+    expect(codeImport._summary).toBeDefined()
+  })
+
+  describe('::processStream', () => {
+    it('should be defined', () => {
+      expect(codeImport.processStream).toBeDefined()
+    })
+
+    it('should call callback when done', (done) => {
+      codeImport._processBatches = jest.fn()
+      codeImport._processBatches.mockReturnValue(Promise.resolve())
+      const myMockCallback = jest.fn(() => {
+        done()
+      })
+      codeImport.processStream('foo', myMockCallback)
+    })
+  })
+
+  describe('::_buildPredicate', () => {
+    it('should be defined', () => {
+      expect(DiscountCodeImport._buildPredicate).toBeDefined()
+    })
+
+    it('should build predicate with codes from array of code objects', () => {
+      const predicate = DiscountCodeImport._buildPredicate(codes)
+      expect(predicate).toMatch(
+        // eslint-disable-next-line max-len
+        'code in ("WILzALjj417", "WILBZ2UYol8", "WIL1EEWHOnY", "WIopQm5d789", "WIopSEF55789")',
+      )
+    })
+  })
+
+  describe('::_processBatches', () => {
+    it('should be defined', () => {
+      expect(codeImport._processBatches).toBeDefined()
+    })
+
+    it('should process list of codes and call `_createOrUpdate`', async () => {
+      const response = { body: { results: [codes[0]] } }
+      codeImport.client.execute = jest.fn(() => Promise.resolve(response))
+      codeImport._createOrUpdate = jest.fn()
+      await codeImport._processBatches(codes)
+      expect(codeImport._createOrUpdate).toHaveBeenCalledTimes(1)
+      expect(codeImport._createOrUpdate).toHaveBeenCalledWith(
+        codes,
+        response.body.results,
+      )
+    })
+  })
+
+  describe('::_createOrUpdate', () => {
+    const existingCodes = codes.slice(0, 2)
+    beforeEach(() => {
+      codeImport._create = jest.fn(() => Promise.resolve())
+      codeImport._update = jest.fn(() => Promise.resolve())
+    })
+
+    it('should be defined', () => {
+      expect(codeImport._createOrUpdate).toBeDefined()
+    })
+
+    it('should call `_update` if code already exists', async () => {
+      await codeImport._createOrUpdate(codes, existingCodes)
+      expect(codeImport._update).toHaveBeenCalledTimes(2)
+      expect(codeImport._update).toHaveBeenCalledWith(
+        codes[0],
+        existingCodes[0],
```

Comment:
for future reference (maybe you can do it here if you want, it's up to you);
this is quite hard to understand, it would be nice to see to give these a name

```js
const [iAmANewCode] = codes
const [iAmAnExistingCode] = existingCodes

expect(codeImport._update).toHaveBeenCalledWith(iAmANewCode, iAmAnExistingCode)
```